Module for regression plots

shizuchanw opened the pull request:
Added residual plot and prediction error plot

edublancas commented on src/sklearn_evaluation/plot/regression.py:
```diff
@@ -0,0 +1,75 @@
+
+
+@SKLearnEvaluationLogger.log(feature='plot')
+def residual(y_true, y_pred, ax=None):
```
can you add a docstring and document the parameters? Look at the other plots as guide

also, please add an inline example. (check out the Examples section in the other plots)

shizuchanw replied:
Thank you! Will do that

edublancas commented on src/sklearn_evaluation/plot/regression.py:
```diff
@@ -0,0 +1,75 @@
+    # x = y_true.reshape(-1,1);
+    # best fit line
+    model = LinearRegression()
+    model.fit(y_true.reshape((-1, 1)), y_pred)
```
why are we fitting a model here? we should leave that to the user and only get the input data. they might want to use another package

shizuchanw replied:
Oh this was the best fit line between the measured values and predicted values, so linear regression shows the difference between the error and the identity line (when measured values == predicted values)

edublancas commented on src/sklearn_evaluation/plot/regression.py:
```diff
@@ -0,0 +1,75 @@
+    if ax is None:
+        ax = plt.gca()
+
+    # x = y_true.reshape(-1,1);
```
remove commented code

edublancas commented on src/sklearn_evaluation/plot/regression.py:
```diff
@@ -0,0 +1,75 @@
+        raise ValueError('y_true and y_pred are needed to plot '
+                         'Residuals Plot')
+
+    # raise error if parameter's dimension != 1?
```
remove commented code

edublancas reviewed (changes requested):
this is a good start :) please check my comments

idomic commented:
@edublancas I don't think this was ready for review.
@shizuchanw please make sure you tag when ready + make sure the CI is passing.

shizuchanw commented:
@idomic @edublancas Can you review this? Thank you!

idomic commented:
@shizuchanw can you resolve the comments above that you already fixed?

idomic commented on tests/test_regression.py:
```diff
@@ -0,0 +1,20 @@
+@image_comparison(baseline_images=['residual'],
+                  extensions=['png'],
+                  remove_text=True)
+def test_residual():
```
I'd also add a test on the parameters (for instance invalid parameter)

idomic reviewed (commented):
Looks good overall, nice job!
Added a small comment.

edublancas reviewed (changes requested):
the code looks good, but it's missing some documentation. can you add a notebook that walks a user through a complete use case? that will help me and users understand why they should use these plots

please create a notebook here: https://github.com/ploomber/sklearn-evaluation/tree/master/docs/source/nbs

and list it here: https://github.com/ploomber/sklearn-evaluation/blob/master/docs/source/index.rst

I'm thinking something with this:
- sample code that trains a regression
- then explain why you'd use the two plots you implemented and how they help someone assess the quality of a model

shizuchanw commented:
> the code looks good, but it's missing some documentation. can you add a notebook that walks a user through a complete use case? that will help me and users understand why they should use these plots
> 
> please create a notebook here: https://github.com/ploomber/sklearn-evaluation/tree/master/docs/source/nbs
> 
> and list it here: https://github.com/ploomber/sklearn-evaluation/blob/master/docs/source/index.rst
> 
> I'm thinking something with this:
> 
> * sample code that trains a regression
> * then explain why you'd use the two plots you implemented and how they help someone assess the quality of a model

Sounds good! Will be working on the notebook.

For the sample code in the notebook, is it ok if I use something similar to what I added in the examples directory?

edublancas commented:
> For the sample code in the notebook, is it ok if I use something similar to what I added in the examples directory?

sure, you can use something similar. the main difference is that the example is meant to be a quick thing that users can copy paste. while the notebook is meant to provide a full explanation for someone who wants to learn the concepts

shizuchanw commented:
I added the notebook into /nbs and edited user_guide.rst. I built the doc locally and it looked fine. However, when I try to commit, the notebook I added is not included in the stashed changes, and I checked gitignore and didn't see anything that'll cause this. Has this happened before?

edublancas commented:
were you able to solve the notebook issue?

Please let me know when this is ready for review

edublancas commented on examples/residual.py:
```diff
@@ -0,0 +1,21 @@
+y_true = y_test
+
+plot.residual(y_true, y_pred)
+plt.show()
```
looks like this example isn't showing much. only a few dots appear in the plot. `make_classification` is to generate sample data for a classifier:

<img width="633" alt="image" src="https://user-images.githubusercontent.com/989250/201981587-f39c0ac9-1e03-4a20-b227-5a05f5e7e73c.png">

try using `make_regression`: https://scikit-learn.org/stable/modules/generated/sklearn.datasets.make_regression.html

shizuchanw replied:
Ohhh sorry about this :( I changed it in my new version but I don't know why it was not included in the commit. I will fix this

edublancas commented on examples/prediction_error.py:
```diff
@@ -0,0 +1,21 @@
+
+from sklearn_evaluation import plot
+
+data = datasets.make_classification(200, 10, n_informative=5, class_sep=0.65)
```
use `make_regression`, since `make_classification` will generate data for a classification problem

edublancas commented on examples/prediction_error.py:
```diff
@@ -0,0 +1,21 @@
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
+
+est = RandomForestClassifier()
```
change to a regression model

edublancas commented on examples/residual.py:
```diff
@@ -0,0 +1,21 @@
+
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)
+
+est = RandomForestClassifier()
```
change to a regression model

edublancas commented on src/sklearn_evaluation/plot/regression.py:
```diff
@@ -0,0 +1,122 @@
+
+@SKLearnEvaluationLogger.log(feature='plot')
+def residual(y_true, y_pred, ax=None):
+    """
```
rename to `residuals`

edublancas commented on src/sklearn_evaluation/plot/regression.py:
```diff
@@ -0,0 +1,122 @@
+    ax.set_ylabel(ylabel)
+
+def _check_parameter_validity(y_true, y_pred):
+    if any((val is None for val in (y_true, y_pred))):
```
this has "residuals plot" harcoded, but it's also called in the prediction error plot, please add a parameter to control for the plot name

edublancas commented on src/sklearn_evaluation/plot/regression.py:
```diff
@@ -0,0 +1,122 @@
+@SKLearnEvaluationLogger.log(feature='plot')
+def prediction_error(y_true,
+                     y_pred,
+                     model=LinearRegression(),
```
update this to `model=None` since it's a bad practice to have objects as defaults. in the function you can do:

```python
model = model or LinearRegression()
```

this way, the `model` variable will take the user provided value, and if None, it'll take the `LinearRegression()`

edublancas reviewed (changes requested):
a few notes on the notebook. the prediction error plot says "residuals plot" in the title:

<img width="812" alt="image" src="https://user-images.githubusercontent.com/989250/201982892-86143b55-44ea-4dd3-b41d-38c3d62eb9a2.png">

also, the predictions plot says "y_predicted and y_measured", but that doesn't match the function's signature.

shizuchanw commented:
@edublancas Hi Eduardo, I made changes based on your comments. Do you mind checking it again? Thanks!

edublancas commented:
Awesome work! 🎉
